// @sdetsanjay/autoheal-locator/dist/models/LocatorType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocatorTypeDisplayNames = exports.LocatorType = void 0;
/**
 * Types of locators supported
 */
var LocatorType;
(function (LocatorType) {
    LocatorType["CSS_SELECTOR"] = "CSS_SELECTOR";
    LocatorType["XPATH"] = "XPATH";
    LocatorType["ID"] = "ID";
    LocatorType["NAME"] = "NAME";
    LocatorType["CLASS_NAME"] = "CLASS_NAME";
    LocatorType["TAG_NAME"] = "TAG_NAME";
    LocatorType["LINK_TEXT"] = "LINK_TEXT";
    LocatorType["PARTIAL_LINK_TEXT"] = "PARTIAL_LINK_TEXT";
    // Playwright-specific
    LocatorType["GET_BY_ROLE"] = "GET_BY_ROLE";
    LocatorType["GET_BY_LABEL"] = "GET_BY_LABEL";
    LocatorType["GET_BY_PLACEHOLDER"] = "GET_BY_PLACEHOLDER";
    LocatorType["GET_BY_TEXT"] = "GET_BY_TEXT";
    LocatorType["GET_BY_ALT_TEXT"] = "GET_BY_ALT_TEXT";
    LocatorType["GET_BY_TITLE"] = "GET_BY_TITLE";
    LocatorType["GET_BY_TEST_ID"] = "GET_BY_TEST_ID";
})(LocatorType || (exports.LocatorType = LocatorType = {}));
/**
 * Display names for locator types
 */
exports.LocatorTypeDisplayNames = {
    [LocatorType.CSS_SELECTOR]: 'CSS Selector',
    [LocatorType.XPATH]: 'XPath',
    [LocatorType.ID]: 'ID',
    [LocatorType.NAME]: 'Name',
    [LocatorType.CLASS_NAME]: 'Class Name',
    [LocatorType.TAG_NAME]: 'Tag Name',
    [LocatorType.LINK_TEXT]: 'Link Text',
    [LocatorType.PARTIAL_LINK_TEXT]: 'Partial Link Text',
    [LocatorType.GET_BY_ROLE]: 'getByRole',
    [LocatorType.GET_BY_LABEL]: 'getByLabel',
    [LocatorType.GET_BY_PLACEHOLDER]: 'getByPlaceholder',
    [LocatorType.GET_BY_TEXT]: 'getByText',
    [LocatorType.GET_BY_ALT_TEXT]: 'getByAltText',
    [LocatorType.GET_BY_TITLE]: 'getByTitle',
    [LocatorType.GET_BY_TEST_ID]: 'getByTestId',
};
