// @sdetsanjay/autoheal-locator/dist/index.js
"use strict";
/**
 * AutoHeal Locator for JavaScript/TypeScript
 *
 * AI-powered test automation framework that automatically heals broken locators
 * for Playwright and Selenium WebDriver
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AIProvider = exports.defaultLocatorOptions = exports.CachedSelector = exports.LocatorTypeDisplayNames = exports.LocatorType = exports.LocatorStrategy = exports.AutomationFramework = exports.AutoHealReporter = exports.loadConfiguration = exports.CacheType = exports.ExecutionStrategy = exports.FileCache = exports.MemoryCache = exports.GrokService = exports.DeepSeekService = exports.AnthropicService = exports.OpenAIService = exports.GeminiAIService = exports.SeleniumAdapter = exports.PlaywrightAdapter = exports.AutoHealLocatorBuilder = exports.AutoHealLocator = void 0;
// Main API
var AutoHealLocator_1 = require("./core/AutoHealLocator");
Object.defineProperty(exports, "AutoHealLocator", { enumerable: true, get: function () { return AutoHealLocator_1.AutoHealLocator; } });
Object.defineProperty(exports, "AutoHealLocatorBuilder", { enumerable: true, get: function () { return AutoHealLocator_1.AutoHealLocatorBuilder; } });
// Adapters
var PlaywrightAdapter_1 = require("./adapters/PlaywrightAdapter");
Object.defineProperty(exports, "PlaywrightAdapter", { enumerable: true, get: function () { return PlaywrightAdapter_1.PlaywrightAdapter; } });
var SeleniumAdapter_1 = require("./adapters/SeleniumAdapter");
Object.defineProperty(exports, "SeleniumAdapter", { enumerable: true, get: function () { return SeleniumAdapter_1.SeleniumAdapter; } });
// AI Services
var GeminiAIService_1 = require("./ai/GeminiAIService");
Object.defineProperty(exports, "GeminiAIService", { enumerable: true, get: function () { return GeminiAIService_1.GeminiAIService; } });
var OpenAIService_1 = require("./ai/OpenAIService");
Object.defineProperty(exports, "OpenAIService", { enumerable: true, get: function () { return OpenAIService_1.OpenAIService; } });
var AnthropicService_1 = require("./ai/AnthropicService");
Object.defineProperty(exports, "AnthropicService", { enumerable: true, get: function () { return AnthropicService_1.AnthropicService; } });
var DeepSeekService_1 = require("./ai/DeepSeekService");
Object.defineProperty(exports, "DeepSeekService", { enumerable: true, get: function () { return DeepSeekService_1.DeepSeekService; } });
var GrokService_1 = require("./ai/GrokService");
Object.defineProperty(exports, "GrokService", { enumerable: true, get: function () { return GrokService_1.GrokService; } });
// Cache implementations
var MemoryCache_1 = require("./cache/MemoryCache");
Object.defineProperty(exports, "MemoryCache", { enumerable: true, get: function () { return MemoryCache_1.MemoryCache; } });
var FileCache_1 = require("./cache/FileCache");
Object.defineProperty(exports, "FileCache", { enumerable: true, get: function () { return FileCache_1.FileCache; } });
// Configuration
var AutoHealConfiguration_1 = require("./config/AutoHealConfiguration");
Object.defineProperty(exports, "ExecutionStrategy", { enumerable: true, get: function () { return AutoHealConfiguration_1.ExecutionStrategy; } });
Object.defineProperty(exports, "CacheType", { enumerable: true, get: function () { return AutoHealConfiguration_1.CacheType; } });
Object.defineProperty(exports, "loadConfiguration", { enumerable: true, get: function () { return AutoHealConfiguration_1.loadConfiguration; } });
// Reporting
var AutoHealReporter_1 = require("./reporting/AutoHealReporter");
Object.defineProperty(exports, "AutoHealReporter", { enumerable: true, get: function () { return AutoHealReporter_1.AutoHealReporter; } });
// Models
var AutomationFramework_1 = require("./models/AutomationFramework");
Object.defineProperty(exports, "AutomationFramework", { enumerable: true, get: function () { return AutomationFramework_1.AutomationFramework; } });
var LocatorStrategy_1 = require("./models/LocatorStrategy");
Object.defineProperty(exports, "LocatorStrategy", { enumerable: true, get: function () { return LocatorStrategy_1.LocatorStrategy; } });
var LocatorType_1 = require("./models/LocatorType");
Object.defineProperty(exports, "LocatorType", { enumerable: true, get: function () { return LocatorType_1.LocatorType; } });
Object.defineProperty(exports, "LocatorTypeDisplayNames", { enumerable: true, get: function () { return LocatorType_1.LocatorTypeDisplayNames; } });
var CachedSelector_1 = require("./models/CachedSelector");
Object.defineProperty(exports, "CachedSelector", { enumerable: true, get: function () { return CachedSelector_1.CachedSelector; } });
var LocatorRequest_1 = require("./models/LocatorRequest");
Object.defineProperty(exports, "defaultLocatorOptions", { enumerable: true, get: function () { return LocatorRequest_1.defaultLocatorOptions; } });
var AIProvider_1 = require("./models/AIProvider");
Object.defineProperty(exports, "AIProvider", { enumerable: true, get: function () { return AIProvider_1.AIProvider; } });
