// @sdetsanjay/autoheal-locator/dist/cache/FileCache.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileCache = void 0;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const CachedSelector_1 = require("../models/CachedSelector");
/**
 * Persistent file-based cache implementation
 */
class FileCache {
    constructor(cacheDirectory = './autoheal-cache', maxSize = 10000, ttlMs = 24 * 60 * 60 * 1000) {
        this.hitCount = 0;
        this.missCount = 0;
        this.evictionCount = 0;
        this.cacheDirectory = cacheDirectory;
        this.cacheFile = path.join(cacheDirectory, 'selectors.json');
        this.maxSize = maxSize;
        this.ttlMs = ttlMs;
        this.cache = new Map();
        this.ensureCacheDirectory();
        this.loadFromDisk();
    }
    ensureCacheDirectory() {
        if (!fs.existsSync(this.cacheDirectory)) {
            fs.mkdirSync(this.cacheDirectory, { recursive: true });
        }
    }
    loadFromDisk() {
        if (fs.existsSync(this.cacheFile)) {
            try {
                const data = fs.readFileSync(this.cacheFile, 'utf-8');
                const parsed = JSON.parse(data);
                let expiredCount = 0;
                const now = Date.now();
                for (const [key, value] of Object.entries(parsed)) {
                    const typedValue = value;
                    // Check if expired
                    if (now - typedValue.timestamp > this.ttlMs) {
                        expiredCount++;
                        continue;
                    }
                    const cached = Object.assign(new CachedSelector_1.CachedSelector(typedValue.selector), typedValue);
                    this.cache.set(key, cached);
                }
                console.log(`[FILE-CACHE] Loaded ${this.cache.size} entries from cache file, ${expiredCount} expired entries skipped`);
                console.log(`[FILE-CACHE] Initialized persistent file cache at: ${this.cacheDirectory}`);
            }
            catch (error) {
                console.error('[FILE-CACHE] Failed to load cache from disk:', error);
            }
        }
        else {
            console.log(`[FILE-CACHE] No existing cache file found, initializing empty cache at: ${this.cacheDirectory}`);
        }
    }
    saveToDisk() {
        try {
            this.ensureCacheDirectory();
            const data = {};
            for (const [key, value] of this.cache.entries()) {
                data[key] = value;
            }
            fs.writeFileSync(this.cacheFile, JSON.stringify(data, null, 2), 'utf-8');
        }
        catch (error) {
            console.error('Failed to save cache to disk:', error);
        }
    }
    get(key) {
        const value = this.cache.get(key);
        if (value) {
            // Check if expired
            if (Date.now() - value.timestamp > this.ttlMs) {
                this.cache.delete(key);
                this.missCount++;
                console.log(`[FILE-CACHE] Cache EXPIRED: ${key}`);
                return undefined;
            }
            this.hitCount++;
            console.log(`[FILE-CACHE] Cache HIT: ${key}`);
            return value;
        }
        else {
            this.missCount++;
            console.log(`[FILE-CACHE] Cache MISS: ${key}`);
            return undefined;
        }
    }
    put(key, selector) {
        // Evict oldest entries if cache is full
        if (this.cache.size >= this.maxSize) {
            const firstKey = this.cache.keys().next().value;
            if (firstKey) {
                this.cache.delete(firstKey);
                this.evictionCount++;
            }
        }
        this.cache.set(key, selector);
        // Calculate expiry time
        const expiryHours = Math.round(this.ttlMs / (60 * 60 * 1000));
        console.log(`[FILE-CACHE] Cache STORED: ${key} (expires in ${expiryHours} hours)`);
        this.saveToDisk();
    }
    remove(key) {
        const result = this.cache.delete(key);
        if (result) {
            this.saveToDisk();
        }
        return result;
    }
    clearAll() {
        this.cache.clear();
        this.hitCount = 0;
        this.missCount = 0;
        this.evictionCount = 0;
        this.saveToDisk();
    }
    size() {
        return this.cache.size;
    }
    updateSuccess(key, success) {
        const cached = this.cache.get(key);
        if (cached) {
            cached.updateSuccess(success);
            this.cache.set(key, cached);
            this.saveToDisk();
        }
    }
    evictExpired() {
        const now = Date.now();
        const keysToDelete = [];
        for (const [key, value] of this.cache.entries()) {
            if (now - value.timestamp > this.ttlMs) {
                keysToDelete.push(key);
            }
        }
        for (const key of keysToDelete) {
            this.cache.delete(key);
            this.evictionCount++;
        }
        if (keysToDelete.length > 0) {
            this.saveToDisk();
            console.log(`Evicted ${keysToDelete.length} expired cache entries`);
        }
    }
    getMetrics() {
        const total = this.hitCount + this.missCount;
        return {
            hitCount: this.hitCount,
            missCount: this.missCount,
            hitRate: total === 0 ? 0 : this.hitCount / total,
            totalEntries: this.cache.size,
            evictionCount: this.evictionCount,
        };
    }
}
exports.FileCache = FileCache;
