// @sdetsanjay/autoheal-locator/dist/config/AutoHealConfiguration.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_CONFIG = exports.CacheType = exports.ExecutionStrategy = void 0;
exports.loadConfiguration = loadConfiguration;
const dotenv = __importStar(require("dotenv"));
const path = __importStar(require("path"));
const fs = __importStar(require("fs"));
const AIProvider_1 = require("../models/AIProvider");
dotenv.config();
/**
 * Execution strategy for element healing
 *
 * Determines the order and approach for AI-powered element location:
 *
 * - **SMART_SEQUENTIAL** (Default):
 *   1. Try DOM analysis first (fast & cost-effective)
 *   2. If DOM fails to find element → Automatically fallback to Visual analysis
 *   3. Best for most use cases (balances speed, cost, and reliability)
 *
 * - **VISUAL_FIRST**:
 *   1. Try Visual analysis first (better for complex/dynamic UIs)
 *   2. If Visual fails to find element → Automatically fallback to DOM analysis
 *   3. Higher cost but better for visually-driven element location
 *
 * - **DOM_ONLY**:
 *   1. Only use DOM analysis (no visual fallback)
 *   2. Fastest and cheapest option
 *   3. Use when elements are reliably accessible via DOM
 *
 * - **PARALLEL**:
 *   1. Run DOM and Visual analysis concurrently (future implementation)
 *   2. Currently uses VISUAL analysis only
 *   3. Highest cost but fastest when both methods are needed
 *
 * - **SEQUENTIAL**:
 *   1. Legacy/deprecated - same as SMART_SEQUENTIAL
 *   2. Use SMART_SEQUENTIAL instead
 */
var ExecutionStrategy;
(function (ExecutionStrategy) {
    ExecutionStrategy["DOM_ONLY"] = "DOM_ONLY";
    ExecutionStrategy["SMART_SEQUENTIAL"] = "SMART_SEQUENTIAL";
    ExecutionStrategy["PARALLEL"] = "PARALLEL";
    ExecutionStrategy["VISUAL_FIRST"] = "VISUAL_FIRST";
    ExecutionStrategy["SEQUENTIAL"] = "SEQUENTIAL";
})(ExecutionStrategy || (exports.ExecutionStrategy = ExecutionStrategy = {}));
/**
 * Cache type configuration
 */
var CacheType;
(function (CacheType) {
    CacheType["MEMORY"] = "MEMORY";
    CacheType["PERSISTENT_FILE"] = "PERSISTENT_FILE";
})(CacheType || (exports.CacheType = CacheType = {}));
/**
 * Default configuration values
 */
exports.DEFAULT_CONFIG = {
    ai: {
        provider: AIProvider_1.AIProvider.GOOGLE_GEMINI,
        model: 'gemini-2.0-flash-exp',
        timeout: 30000,
        maxRetries: 3,
        visualAnalysisEnabled: true,
    },
    cache: {
        type: CacheType.PERSISTENT_FILE,
        maxSize: 10000,
        expireAfterWriteMs: 24 * 60 * 60 * 1000, // 24 hours
        cacheDirectory: './autoheal-cache',
    },
    performance: {
        executionStrategy: ExecutionStrategy.SMART_SEQUENTIAL,
        threadPoolSize: 4,
        elementTimeoutMs: 30000,
        enableMetrics: true,
    },
    reporting: {
        enabled: true,
        generateHTML: true,
        generateJSON: true,
        generateText: false,
        consoleLogging: true,
        outputDirectory: './autoheal-reports',
    },
};
/**
 * Load configuration from environment variables and files
 */
function loadConfiguration(customConfig) {
    const envConfig = loadFromEnvironment();
    const fileConfig = loadFromConfigFile();
    // Merge: custom > file > env > defaults
    return deepMerge(exports.DEFAULT_CONFIG, envConfig, fileConfig, customConfig || {});
}
/**
 * Load configuration from environment variables
 */
function loadFromEnvironment() {
    const config = {};
    // AI Configuration
    if (process.env.AUTOHEAL_AI_PROVIDER) {
        config.ai = {
            provider: process.env.AUTOHEAL_AI_PROVIDER,
            apiKey: getAPIKeyFromEnv(process.env.AUTOHEAL_AI_PROVIDER),
            model: process.env.AUTOHEAL_AI_MODEL,
            timeout: parseInt(process.env.AUTOHEAL_AI_TIMEOUT || '30000'),
            maxRetries: parseInt(process.env.AUTOHEAL_AI_MAX_RETRIES || '3'),
            visualAnalysisEnabled: process.env.AUTOHEAL_AI_VISUAL_ENABLED !== 'false',
        };
    }
    // Cache Configuration
    if (process.env.AUTOHEAL_CACHE_TYPE) {
        config.cache = {
            type: process.env.AUTOHEAL_CACHE_TYPE,
            maxSize: parseInt(process.env.AUTOHEAL_CACHE_MAX_SIZE || '10000'),
            expireAfterWriteMs: parseInt(process.env.AUTOHEAL_CACHE_EXPIRE_AFTER_WRITE || '86400000'),
            cacheDirectory: process.env.AUTOHEAL_CACHE_DIRECTORY,
        };
    }
    // Performance Configuration
    if (process.env.AUTOHEAL_EXECUTION_STRATEGY) {
        config.performance = {
            executionStrategy: process.env.AUTOHEAL_EXECUTION_STRATEGY,
            threadPoolSize: parseInt(process.env.AUTOHEAL_THREAD_POOL_SIZE || '4'),
            elementTimeoutMs: parseInt(process.env.AUTOHEAL_ELEMENT_TIMEOUT || '30000'),
            enableMetrics: process.env.AUTOHEAL_ENABLE_METRICS !== 'false',
        };
    }
    // Reporting Configuration
    if (process.env.AUTOHEAL_REPORTING_ENABLED !== undefined) {
        config.reporting = {
            enabled: process.env.AUTOHEAL_REPORTING_ENABLED === 'true',
            generateHTML: process.env.AUTOHEAL_REPORTING_HTML !== 'false',
            generateJSON: process.env.AUTOHEAL_REPORTING_JSON !== 'false',
            generateText: process.env.AUTOHEAL_REPORTING_TEXT === 'true',
            consoleLogging: process.env.AUTOHEAL_REPORTING_CONSOLE !== 'false',
            outputDirectory: process.env.AUTOHEAL_REPORTING_OUTPUT_DIR || './autoheal-reports',
        };
    }
    return config;
}
/**
 * Get API key from environment based on provider
 */
function getAPIKeyFromEnv(provider) {
    switch (provider) {
        case AIProvider_1.AIProvider.GOOGLE_GEMINI:
            return process.env.GEMINI_API_KEY;
        case AIProvider_1.AIProvider.OPENAI:
            return process.env.OPENAI_API_KEY;
        case AIProvider_1.AIProvider.ANTHROPIC:
            return process.env.ANTHROPIC_API_KEY;
        case AIProvider_1.AIProvider.DEEPSEEK:
            return process.env.DEEPSEEK_API_KEY;
        case AIProvider_1.AIProvider.GROK:
            return process.env.GROK_API_KEY;
        case AIProvider_1.AIProvider.GROQ:
            return process.env.GROQ_API_KEY;
        default:
            return undefined;
    }
}
/**
 * Load configuration from .autohealrc.json file
 */
function loadFromConfigFile() {
    const configPaths = [
        path.join(process.cwd(), '.autohealrc.json'),
        path.join(process.cwd(), 'autoheal.config.json'),
    ];
    for (const configPath of configPaths) {
        if (fs.existsSync(configPath)) {
            try {
                const fileContent = fs.readFileSync(configPath, 'utf-8');
                return JSON.parse(fileContent);
            }
            catch (error) {
                console.warn(`Failed to load config from ${configPath}:`, error);
            }
        }
    }
    return {};
}
/**
 * Deep merge configuration objects
 */
function deepMerge(...objects) {
    return objects.reduce((prev, obj) => {
        if (!obj)
            return prev;
        Object.keys(obj).forEach((key) => {
            const prevValue = prev[key];
            const objValue = obj[key];
            if (Array.isArray(prevValue) && Array.isArray(objValue)) {
                prev[key] = objValue;
            }
            else if (typeof prevValue === 'object' && typeof objValue === 'object') {
                prev[key] = deepMerge(prevValue, objValue);
            }
            else if (objValue !== undefined) {
                prev[key] = objValue;
            }
        });
        return prev;
    }, {});
}
